Automating Testnet and Futurenet Reset Data in Stellar

KodeSage opened the pull request:
Bounty PR for Issue #608

KodeSage commented:
@briwylde08 this is ready for review.

briwylde08 commented:
Thank you for the submission! We'll get this assigned and reviewed.

ElliotFriend reviewed (changes requested):
Hi, there @KodeSage! Thanks a bunch for this submission! I think you've got a pretty good start here! You've done a really good job setting up the context of network resets and the benefits. I've added some small notes and suggestions in a couple places.

I also want to request some more general changes to the script you've put together.

- The `checkNetworkStatus` function probably doesn't really need to be its own function. Additionally, you are probably better served using the `server.getHealth()` method, as that would tell you some relevant details if Testnet weren't read yet.
- The [docs page you link to](https://developers.stellar.org/docs/learn/fundamentals/networks#test-data-automation) mentions a few different scenarios that a script like this might cover. I think we'd do well to cover at least _most_ of those listed scenarios. Maybe asset issuance, liquidity pools, and contract deployments? Your script could use a function to issue the asset, one for creating a liquidity pool, and another for deploying/testing/etc. a smart contract.
- Your `issueAsset` function probably needs a little work. In its current state, it really only makes a payment of the native asset, whereas truly "issuing an asset" would involve creating a trustline on the receiving account, and a payment from the issuing account. Having a predictable asset issued on Testnet is probably one of the more important things we'll want this script to cover, as it's the "bread and butter" kind of data that a lot of folks will need to re-create.

How does that all sound to you? Would love to know your thoughts. If you've got any questions, or need any help, don't hesitate to reach out.

Thanks again!

KodeSage commented:
Hi @ElliotFriend , Thank you so much for the reviews and feedbacks 😁. 
I have updated and made some changes, Including Adding a `CreateLiquidityPool` and `DeployContract` scripts.
also I updated the I`ssueAsset` script too and corrected the nitpicks you pointed out.

kindly please review again, Thank you.

ElliotFriend reviewed (commented):
Hi, there! This is coming along really well! I've got a few notes for you, and a bit of refactoring needs to be done. But, I think once you have these things taken care of we'll be pretty much ready to go.

Thanks!!

KodeSage commented:
Alright , Thank you for this Feedbacks @ElliotFriend , I will look into now and make necessary changes.

KodeSage commented:
Hello @ElliotFriend , I made neccssary corrections and updated the code walk flows too.

ElliotFriend commented on docs/build/guides/basics/automate-reset-data.mdx:
```diff
@@ -0,0 +1,460 @@
+        case "FAILED":
+          throw tmpStatus;
+        case "NOT_FOUND":
+          await sleep(1000); // Increased sleep time
```
```suggestion
          await sleep(1000); // Wait for one minute
```

Let's make these comments a little more instructive

ElliotFriend commented on docs/build/guides/basics/automate-reset-data.mdx:
```diff
@@ -0,0 +1,460 @@
+
+    let status;
+    let attempts = 0;
+    while (attempts++ < 20) {
+      // Increased max attempts
```
```suggestion
    // A maximum of 20 polls of transaction status
    while (attempts++ < 20) {
```

ElliotFriend commented on docs/build/guides/basics/automate-reset-data.mdx:
```diff
@@ -0,0 +1,460 @@
+    }
+
+    if (attempts >= 20 || !status) {
+      // Updated max attempts check
```
```suggestion
```

ElliotFriend commented on docs/build/guides/basics/automate-reset-data.mdx:
```diff
@@ -0,0 +1,460 @@
+function sleep(ms) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+const getTxsReports = async (feedback, message, server) => {
+  try {
+    const result = await yeetTx(feedback, server);
+    console.log(`${message}: Transaction succeeded`, result);
+    return result;
+  } catch (error) {
+    console.error(`${message}: Transaction failed`, error);
+  }
```
I don't think this function really adds much for us. I do like the idea of logging this status `message` along the way, but I think you're probably better off modifying the `yeetTx()` function itself.

ElliotFriend commented on docs/build/guides/basics/automate-reset-data.mdx:
```diff
@@ -0,0 +1,386 @@
+async function createAccount(networkURL) {
+  try {
+    // Generate a keypair
+    const pair = StellarSdk.Keypair.random();
```
This hasn't been addressed yet.

ElliotFriend commented on docs/build/guides/basics/automate-reset-data.mdx:
```diff
@@ -0,0 +1,460 @@
+    const nativeAsset = StellarSdk.Asset.native();
+    await createLiquidityPool(
+      server,
+      accountOne,
```
We should probably use `accountTwo` here (the non-issuer account), so it's a little more realistic.

ElliotFriend commented on docs/build/guides/basics/automate-reset-data.mdx:
```diff
@@ -0,0 +1,460 @@
+        StellarSdk.Operation.liquidityPoolDeposit({
+          liquidityPoolId: poolId,
+          maxAmountA: "1000", // Amount of asset A to deposit
+          maxAmountB: "1000", // Amount of asset B to deposit
```
Since you're going to use `accountTwo` for these LP operations, you will need to modify `maxAmountB`. Your payment operation from earlier only sent 800 to `accountTwo`, so this operation would fail. Either the payment or this deposit op should be changed.

ElliotFriend commented on docs/build/guides/basics/automate-reset-data.mdx:
```diff
@@ -0,0 +1,460 @@
+    // FOR FUTERENET - https://friendbot-futurenet.stellar.org
+    // FOR TESTNET - https://friendbot.stellar.org
+
+    const accountOne = await createAccount(networkURL);
+    const accountTwo = await createAccount(networkURL);
```
I mentioned in my previous review, that (at least) `accountOne` **needs** to not be created by `Keypair.random()`. We're trying to make this a useful script for people who need to setup predictable things after a network reset. Among the most important things that should be "predictable" is an asset's issuer address.

ElliotFriend commented on docs/build/guides/basics/automate-reset-data.mdx:
```diff
@@ -0,0 +1,460 @@
+    const contractWasmFilePath =
+      "./target/wasm32-unknown-unknown/release/hello_world.wasm"; // This is Wasmfile for an Hello World Contract.
+    const salt =
+      "3157dca02e4105a749bd88d0ea429cde67b399aaa99dc646edccd74d5692a857"; // example salt
```
Feel free to drop in your own "easter egg" into the `salt` here. For example `KodeSage was here!` in hex format would be `4b6f64655361676520776173206865726521`. Obviously, you don't have to, but I'm a dork and I like to do stuff like that 🤣

ElliotFriend reviewed (changes requested):
Hey, there! It's looking really good, very close. I've left one last round of comments, and I think we'll be good to go. There was one from the previous review that got missed, too. Thanks!!

KodeSage commented:
HI @ElliotFriend , GM!

I have updated the scripts.

For the `createAccount` I made is flexible for users to restore their account or create an account.

ElliotFriend reviewed (approved):
Lookin' really good, @KodeSage!! Thanks a bunch for the hard work you've put into the script, and the instructions as well! Well done!

briwylde08 commented:
Hi @KodeSage! Thank you again for another great submission! We'll get this merged and take the award conversation back to the corresponding issue.
